OpenRA.Mods.Common/Activities/Rearm.cs

```diff
@@ -8,53 +8,64 @@
  */
 #endregion
 
+using System.Collections.Generic;
 using System.Linq;
 using OpenRA.Activities;
 using OpenRA.Mods.Common.Traits;
-using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Activities
 {
 	public class Rearm : Activity
 	{
-		readonly LimitedAmmo limitedAmmo;
-		int ticksPerPip = 25 * 2;
-		int remainingTicks = 25 * 2;
-		string sound = null;
+		readonly IEnumerable<AmmoPool> ammoPools;
+		readonly Dictionary<AmmoPool, int> ammoPoolsReloadTimes;
 
-		public Rearm(Actor self, string sound = null)
+		public Rearm(Actor self)
 		{
-			limitedAmmo = self.TraitOrDefault<LimitedAmmo>();
-			if (limitedAmmo != null)
-				ticksPerPip = limitedAmmo.ReloadTimePerAmmo();
-			remainingTicks = ticksPerPip;
-			this.sound = sound;
+			ammoPools = self.TraitsImplementing<AmmoPool>();
+
+			if (ammoPools == null)
+				return;
+
+			ammoPoolsReloadTimes = ammoPools.ToDictionary(x => x, y => y.Info.ReloadTicks);
 		}
 
 		public override Activity Tick(Actor self)
 		{
-			if (IsCanceled || limitedAmmo == null)
+			if (IsCanceled || ammoPoolsReloadTimes == null)
 				return NextActivity;
 
-			if (--remainingTicks == 0)
+			var needsReloading = false;
+
+			foreach (var pool in ammoPools)
 			{
-				var hostBuilding = self.World.ActorMap.GetUnitsAt(self.Location)
-					.FirstOrDefault(a => a.HasTrait<RenderBuilding>());
+				if (pool.FullAmmo())
+					continue;
+
+				needsReloading = true;
+
+			    if (--ammoPoolsReloadTimes[pool] > 0)
+			        continue;
+
+				// HACK to check if we are on the helipad/airfield/etc.
+				var hostBuilding = self.World.ActorMap.GetUnitsAt(self.Location).FirstOrDefault(a => a.HasTrait<RenderBuilding>());
 
 				if (hostBuilding == null || !hostBuilding.IsInWorld)
 					return NextActivity;
```

Comment:
This may have been a very bad call on my part :/
The performance hit from doing this every tick instead of when we need to give 1 more ammo is probably higher than it was before you moved this :(
Which means that this hack will need to be resolved fast.
I started working on proper docking for repairing yesterday. So far it works a bit better then what's on bleed, so I should be able to do it for rearming soon (hopefully).